com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs

```diff
@@ -1,21 +1,478 @@
 using System.Collections.Generic;
+using System.Linq;
 using UnityEngine;
 using UnityEditor.Experimental.GraphView;
+using UnityEngine.UIElements;
+using System;
+using UnityEditor.Graphing;
+using UnityEditor.ShaderGraph.Drawing.Views;
+using UnityEditor.ShaderGraph.Internal;
+using UnityEngine.Assertions;
+using GraphDataStore = UnityEditor.ShaderGraph.DataStore<UnityEditor.ShaderGraph.GraphData>;
+using BlackboardItem = UnityEditor.ShaderGraph.Internal.ShaderInput;
 
-namespace UnityEditor.ShaderGraph.Drawing.Views.Blackboard
+namespace UnityEditor.ShaderGraph.Drawing
 {
-    class BlackboardController : SGViewController<BlackboardViewModel>
+    class AddShaderInputAction : IGraphDataAction
     {
-        IList<SGBlackboardSection> m_BlackboardSections;
+        void AddShaderInput(GraphData graphData)
+        {
+            Assert.IsNotNull(graphData, "GraphData is null while carrying out AddShaderInputAction");
+            // If type property is valid, create instance of that type
+            if (blackboardItemType != null && blackboardItemType.IsSubclassOf(typeof(BlackboardItem)))
+                shaderInputReference = (BlackboardItem)Activator.CreateInstance(blackboardItemType, true);
+            // If type is null a direct override object must have been provided or else we are in an error-state
+            else if (shaderInputReference == null)
+            {
+                Debug.Log("ERROR: BlackboardController: Unable to complete Add Shader Input action.");
+                return;
+            }
+
+            shaderInputReference.generatePropertyBlock = shaderInputReference.isExposable;
+
+            graphData.owner.RegisterCompleteObjectUndo("Add Shader Input");
+            graphData.AddGraphInput(shaderInputReference);
+        }
+
+        public Action<GraphData> modifyGraphDataAction => AddShaderInput;
+
+        // If this is a subclass of ShaderInput and is not null, then an object of this type is created to add to blackboard
+        public Type blackboardItemType { get; set; }
+
+        // If the type field above is null and this is provided, then it is directly used as the item to add to blackboard
+        public BlackboardItem shaderInputReference { get; set; }
+    }
+
+    class ChangeGraphPathAction : IGraphDataAction
+    {
+        void ChangeGraphPath(GraphData graphData)
+        {
+            if (graphData != null)
+            {
+                graphData.path = NewGraphPath;
+            }
+        }
+
+        public Action<GraphData> modifyGraphDataAction => ChangeGraphPath;
+
+        public string NewGraphPath { get; set; }
+    }
+
+    class CopyShaderInputAction : IGraphDataAction
+    {
+        void CopyShaderInput(GraphData graphData)
+        {
+            Assert.IsNotNull(graphData, "GraphData is null while carrying out CopyShaderInputAction");
+            Assert.IsNotNull(shaderInputToCopy, "ShaderInputToCopy is null while carrying out CopyShaderInputAction");
```

Comment:
I don't think we want to land Asserts in our real code - that'll throw an uncaught exception that will put the graph in a bad state. Unless they're compiled out? I'm not actually sure.

Reply:
Very good catch, negligent of me to leave these in here, have been reading up on standards for assertions in production code, i.e. **DON'T DO IT**

I have wrapped them in an `#if SG_ASSERTIONS` check that we can use to compile those instructions out of production, but we can turn them on during our local dev-work in order to make sure that anything that we do doesn't regress the graph data manipulation.